keras_rs/src/utils/ranking_metrics_utils.py

```diff
@@ -0,0 +1,240 @@
+) -> types.Tensor:
+    """Utility function for getting shuffled indices, with masked indices
+    pushed to the end.
+
+    Args:
+        shape: tuple. The shape of the tensor for which to generate
+            shuffled indices.
+        mask: An optional boolean tensor with the same shape as `shape`.
+            If provided, elements where `mask` is `False` will be placed
+            at the end of the sorted indices. Defaults to `None` (no masking).
+        shuffle_ties: Boolean indicating how to handle ties if multiple elements
+            have the same sorting value (randomly when `shuffle_ties` is True
+            otherwise, order is preserved).
+        seed: Optional integer seed for the random number generator used when
+            `shuffle_ties` is True. Ensures reproducibility. Defaults to None.
+
+    Returns:
+        A tensor of shape `shape` containing shuffled indices.
+    """
+    # If `shuffle_ties` is True, generate random values. Otherwise, generate
+    # zeros so that we get `[0, 1, 2, ...]` as indices on doing `argsort`.
+    if shuffle_ties:
+        shuffle_values = keras.random.uniform(shape, seed=seed, dtype="float32")
+    else:
+        shuffle_values = ops.zeros(shape, dtype="float32")
+
+    # When `mask = False`, increase value by 1 so that those indices are placed
+    # at the end. Note that `shuffle_values` lies in the range `[0, 1)`, so
+    # adding by 1 works out.
+    if mask is not None:
+        shuffle_values = ops.where(
+            mask,
+            shuffle_values,
+            ops.add(shuffle_values, ops.cast(1, dtype="float32")),
+        )
+
+    shuffled_indices = ops.argsort(shuffle_values)
+    return shuffled_indices
+
+
+def sort_by_scores(
+    tensors_to_sort: list[types.Tensor],
+    scores: types.Tensor,
+    mask: Optional[types.Tensor] = None,
+    k: Optional[int] = None,
+    seed: Optional[int] = None,
+    shuffle_ties: bool = True,
+) -> types.Tensor:
+    """
+    Utility function for sorting tensors by scores.
+
+    Args:
+        tensors_to_sort. list of tensors. All tensors are of shape
+            `(batch_size, list_size)`. These tensors are sorted based on
+            `scores`.
+        scores: tensor. Of shape `(batch_size, list_size)`. The scores to sort
+            by.
+        k: int. The number of top-ranked items to consider (the 'k' in 'top-k').
+            If `None`, `list_size` is used.
+        seed: int. Seed for shuffling.
+        shuffle_ties: bool. Whether to randomly shuffle scores before sorting.
+            This is done to break ties.
+
+    Returns:
+        List of sorted tensors (`tensors_to_sort`), sorted using `scores`.
+    """
+    # TODO: Consider exposing `shuffle_ties` to the user.
+    # TODO: Figure out `seed`. How do we propagate it down here from ranking
+    # metric?
+    max_possible_k = ops.shape(scores)[1]
+    if k is None:
+        k = max_possible_k
+    else:
+        k = ops.minimum(k, max_possible_k)
```

Comment:
I would split this in 2, because JAX will want a static int value and TensorFlow may need a dynamic value when `max_possible_k` is a scalar tensor:

```python
    if k is None:
        k = max_possible_k
    elif isinstance(max_possible_k, int):
        k = min(k, max_possible_k)
    else:
        k = ops.minimum(k, max_possible_k)
```